packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.tsx

```diff
@@ -1,427 +1,187 @@
-      !item.name && ('ms-CommandBarItem--noName ' + styles.itemLinkIsNoName),
-      (expandedMenuItemKey === item.key) && ('is-expanded ' + styles.itemLinkIsExpanded)
-    );
-    let hasIcon = !!item.icon || !!item.iconProps;
-
-    return <div className={ css('ms-CommandBarItem', styles.item, item.className) } key={ itemKey } ref={ itemKey }>
-      { (() => {
-        if (isLink) {
-          return <button
-            { ...getNativeProps(item, buttonProperties) }
-            id={ this._id + item.key }
-            className={ className }
-            onClick={ (ev) => this._onItemClick(ev, item) }
-            data-command-key={ index }
-            aria-haspopup={ hasSubmenuItems(item) }
-            aria-expanded={ hasSubmenuItems(item) ? expandedMenuItemKey === item.key : undefined }
-            role='menuitem'
-            aria-label={ item.ariaLabel }
-          >
-            { (hasIcon) ? this._renderIcon(item) : (null) }
-            { (!!item.name) && (
-              <span
-                className={ css('ms-CommandBarItem-commandText', styles.itemCommandText) }
-              >
-                { item.name }
-              </span>
-            ) }
-            { hasSubmenuItems(item) ? (
-              <Icon className={ css('ms-CommandBarItem-chevronDown', styles.itemChevronDown) } iconName='ChevronDown' />
-            ) : (null) }
-          </button>;
-        } else if (item.href) {
-          return <a
-            { ...getNativeProps(item, anchorProperties) }
-            id={ this._id + item.key }
-            className={ className }
-            href={ item.href }
-            data-command-key={ index }
-            aria-haspopup={ hasSubmenuItems(item) }
-            role='menuitem'
-            aria-label={ item.ariaLabel }
-          >
-            { (hasIcon) ? this._renderIcon(item) : (null) }
-            { (!!item.name) && (
-              <span
-                className={ css('ms-CommandBarItem-commandText', styles.itemCommandText) }
-              >
-                { item.name }
-              </span>
-            ) }
-          </a>;
-        } else {
-          return <div
-            { ...getNativeProps(item, divProperties) }
-            id={ this._id + item.key }
-            className={ className }
-            data-command-key={ index }
-            aria-haspopup={ hasSubmenuItems(item) }
-          >
-            { (hasIcon) ? this._renderIcon(item) : (null) }
-            { (!!item.name) && (
-              <span
-                className={ css('ms-CommandBarItem-commandText', styles.itemCommandText) }
-                aria-hidden='true'
-                role='presentation'
-              >
-                { item.name }
-              </span>
-            ) }
-          </div>;
-        }
-      })() }
-    </div>;
+  private computeCacheKey(primaryItems: ICommandBarItemProps[]): string {
```

Comment:
We also need to be very careful here, we should make it clear on the props comments that things with the same key are assumed to have the same horizontal width. We need to make sure people simply aren't assigning indices in an array or something like that since that will be buggy.

Do we have a scenario where a command sometimes appears in the top bar with a label, but after some scaling logic it no longer has a label? If so, we need to make sure the label and no label cases have different cache keys